R/input-utils.R

```diff
@@ -90,21 +90,25 @@ generateOptions <- function(inputId, selected, inline, type = 'checkbox',
   div(class = "shiny-options-group", options)
 }
 
-# Take a vector or list, and convert to named list. Also, if any children are
+# Take a vector/list/factor, and convert to named list. Also, if any children are
 # vectors with length > 1, convert those to list. If the list is unnamed,
 # convert it to a named list with blank names.
-listify <- function(x) {
+listify <- function(x, child = FALSE) {
```

Comment:
I'm finding the logic for this function really hard to follow and I want to make sure that it's well documented so that in the future we understand what's going on here. This function has always been surprisingly difficult to understand -- I've spent over an hour trying to fully understand it, and many hours in the past trying to understand previous versions of it, so clearly something isn't right.

I think this function currently does a mix of different things. This is what I think it does -- please feel free to correct me if this isn't right.

The input can be one of:
* Unnamed vector (usually character, but other types are OK)
* Named vector (usually character, but other types are OK)
* Unnamed factor
* Named factor
* Unnamed list
* Named list

If the input is a list, then each element of the list can be any of those types. However, it doesn't recurse past that; there can't be any grandchild nodes. (Note: I just realized that this function currently allows grandchildren, but the shiny input functions which call this function do not make use of grandchildren.)

For top-level inputs, this function does the following for each type.

* Unnamed vector: Converts to named list where names are same as values.
* Named vector: Converts to named list.
* Unnamed factor: Converts to named list where names are same as values.
* Named factor: Converts to named list.
* Unnamed list: Adds names and recurses. (Note: how does it choose the names if the children aren't single elements?)
 Named list: Recurses.

For children, this function does the following for each type:

* Unnamed vector:
  * If length 1: Converts to char vector (does not add names).
  * If length >1: Converts to named list where names are same as values.
* Named vector: Converts to named list.
* Unnamed factor:
  * If length 1: Converts to char vector (does not add names).
  * If length >1: Converts to named list where names are same as values.
* Named factor: Converts to named list.
* Unnamed list: ??
* Named list: ??

I've put "??" in some places because I haven't figured out exactly what happens

At any rate, I think that all this complexity has led to the bugs that we've had with this function, and we probably should do this a different way, like having multiple functions that each do a simpler thing to the input. (Though if you can describe what's happening in a way that's simple to understand and can put it in a comments, that would also be a reasonable way to deal with this.)

Reply:
To answer your question, named and unnamed list children become named lists (first branch).

In a previous attempt to clarify this function I made it generic, but maybe that wasn't the home run I hoped because type is only one of the three things the function dispatches on. Depth and named-ness are the others. It was also significantly longer.


That grandchildren can't be is a great observation. That simplifies the concept of depth and illuminates an opportunity to throw an error instead of process deeper.